### PR TITLE
Add tests for broadcast vectorization

### PR DESCRIPTION
**Pull request by pchintalapudi**
This tests that basic broadcasting operations (outer/inner products) do end up vectorizing after the LLVM pipeline.

@gbaraldi This will probably block LLVM 15 due to https://discourse.llvm.org/t/newpm-loop-unswitch-regression-from-legacy-pm-to-new-pm/69823 and https://reviews.llvm.org/D124376

### CODE REVIEW

**Comment by gbaraldi** [comment]
What optimization is the new LoopUnswitch lacking? Given that it doesn't optimize on LLVM 16 as well, we might be stuck for a very long while. Can we help it from somewhere else?

**Comment by pchintalapudi** [comment]
I'd suspect that we're generating select instructions for the broadcasting operations, and loop unswitch doesn't have support for unswitching select instructions because it won't reduce code size measurably. Most likely we'll need to patch in optional support for unswitch-on-select upstream.

**Comment by gbaraldi** [comment]
And the old one did support doing that? Or is it caused by something else in the pipeline?

**Comment by pchintalapudi** [comment]
I at least see select instruction handling in line 824 of old LoopUnswitch and a NumSelects statistic, whereas SimpleLoopUnswitch.cpp only has NumBranches and NumSwitches.

**Review by vchuravy — APPROVED** [review]

File: test/llvmpasses/pipeline-o2-broadcast.jl [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+# CHECK: store <[[VEC_FACTOR]] x i32>
+
+# COM: Int64
+# CHECK: @japi1_inner_product
+# CHECK: load <[[VEC_FACTOR:[0-9]+]] x i64>
+# CHECK: mul <[[VEC_FACTOR]] x i64>
+# CHECK: store <[[VEC_FACTOR]] x i64>
+
+function inner_product(R, x, y)
+    R .= x' .* y
```
Comment (chriselrod) [inline test/llvmpasses/pipeline-o2-broadcast.jl:63]: Personally, I'd call this an outer product.

I'd also call `squash` an elementwise product between a matrix and the outer product of two vectors.

**Comment by giordano** [comment]
Quite unexpectedly, this new test fails on A64FX:
```
FAIL: Julia :: pipeline-o2-broadcast.jl (31 of 35)
******************** TEST 'Julia :: pipeline-o2-broadcast.jl' FAILED ********************
Script:
--
: 'RUN: at line 1';   julia --startup-file=no -O2 --check-bounds=auto /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/pipeline-o2-broadcast.jl /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/Outpu
t/pipeline-o2-broadcast.jl.tmp -O && llvm-link -S /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/Output/pipeline-o2-broadcast.jl.tmp/* | FileCheck /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/
pipeline-o2-broadcast.jl
: 'RUN: at line 2';   julia --startup-file=no -O3 --check-bounds=auto /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/pipeline-o2-broadcast.jl /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/Outpu
t/pipeline-o2-broadcast.jl.tmp -O && llvm-link -S /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/Output/pipeline-o2-broadcast.jl.tmp/* | FileCheck /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/
pipeline-o2-broadcast.jl
--
Exit Code: 1

Command Output (stderr):
--
/snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/pipeline-o2-broadcast.jl:10:10: error: CHECK: expected string not found in input
# CHECK: load <[[VEC_FACTOR:[0-9]+]] x float>
         ^
<stdin>:6:50: note: scanning from here
define nonnull {} addrspace(10)* @japi1_prod_v_vT_69({} addrspace(10)* %0, {} addrspace(10)** noalias nocapture noundef readonly %1, i32 %2) #0 !dbg !15 {
                                                 ^
<stdin>:9:55: note: possible intended match here
 %gcframe169.sub = getelementptr inbounds [3 x {} addrspace(10)*], [3 x {} addrspace(10)*]* %gcframe169, i64 0, i64 0
                                                      ^

Input file: <stdin>
Check file: /snx11273/home/ri-mgiordano/repo/julia/test/llvmpasses/pipeline-o2-broadcast.jl

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'llvm-link'
            2: source_filename = "llvm-link"
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
            4: target triple = "aarch64-unknown-linux-gnu"
            5:
            6: define nonnull {} addrspace(10)* @japi1_prod_v_vT_69({} addrspace(10)* %0, {} addrspace(10)** noalias nocapture noundef readonly %1, i32 %2) #0 !dbg !15 {
check:10'0                                                      X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            7: top:
check:10'0     ~~~~~
            8:  %gcframe169 = alloca [3 x {} addrspace(10)*], align 16
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9:  %gcframe169.sub = getelementptr inbounds [3 x {} addrspace(10)*], [3 x {} addrspace(10)*]* %gcframe169, i64 0, i64 0
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'1                                                           ?                                                                possible intended match
           10:  %3 = bitcast [3 x {} addrspace(10)*]* %gcframe169 to i8*
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  call void @llvm.memset.p0i8.i32(i8* noundef nonnull align 16 dereferenceable(24) %3, i8 0, i32 24, i1 false), !tbaa !19
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           12:  %4 = alloca {} addrspace(10)**, align 8
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           13:  store volatile {} addrspace(10)** %1, {} addrspace(10)*** %4, align 8
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  %thread_ptr = call i8* asm "mrs $0, tpidr_el0", "=r"() #5
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
>>>>>>

--
```

**Comment by pchintalapudi** [comment]
Does A64FX have scalable vectors? Iirc the syntax for that changes to n x vscale x type.

**Comment by pchintalapudi** [comment]
@giordano could you try the latest on A64FX to see if it fixes the issue? If not, could you send the output of `@code_llvm prod_v_vT(zeros(0,0), zeros(0), zeros(0))`?

**Pull request merged**